pallets/paratensor/tests/epoch.rs

```
@@ -85,23 +83,380 @@ fn test_10_graph() {
+		// === Register [validator1, validator2, server1, server2]
+		for key in 0..n as u64 {
+			let (nonce, work): (u64, Vec<u8>) = ParatensorModule::create_work_for_block_number( netuid, block_number, key * 1_000_000);
+			assert_ok!(ParatensorModule::register(<<Test as Config>::Origin>::signed(key), netuid, block_number, nonce, work, key, key));
+			ParatensorModule::add_balance_to_coldkey_account( &key, 1 );
+			ParatensorModule::set_stake_for_testing( &key, 1 );
+		}
+		assert_eq!(ParatensorModule::get_subnetwork_n(netuid), n);
+		run_to_block( 1 ); block_number += 1; // run to next block to ensure weights are set on nodes after their registration block
+		// === Set weights [val1->srv1: 2/3, val1->srv2: 1/3, val2->srv1: 2/3, val2->srv2: 1/3, srv1->srv1: 1, srv2->srv2: 1]
+		for uid in 0..(n/2) as u64 {
+			assert_ok!(ParatensorModule::set_weights(Origin::signed(uid), netuid, ((n/2)..n).collect(), vec![ 2 * (u16::MAX / 3), u16::MAX / 3 ]));
+		}
+		for uid in ((n/2) as u64)..n as u64 {
+			assert_ok!(ParatensorModule::set_weights(Origin::signed(uid), netuid, vec![ uid as u16 ], vec![ u16::MAX ])); // server self-weight
+		}
+		if sparse { ParatensorModule::epoch_sparse( netuid, 1_000_000_000, debug ); }
+		else { ParatensorModule::epoch( netuid, 1_000_000_000, debug ); }
+		// === Dereg server2 at uid3 (least emission) + register new key over uid3
+		let new_key: u64 = n as u64; // register a new key while at max capacity, which means the least incentive uid will be deregistered
+		let (nonce, work): (u64, Vec<u8>) = ParatensorModule::create_work_for_block_number( netuid, block_number, 0);
+		assert_ok!(ParatensorModule::register(<<Test as Config>::Origin>::signed(new_key), netuid, block_number, nonce, work, new_key, new_key));
+		let deregistered_uid: u16 = n-1; // since uid=n-1 only recieved 1/3 of weight, it will get pruned first
+		assert_eq!(new_key, ParatensorModule::get_hotkey_for_net_and_neuron(netuid, deregistered_uid));
```

@SaMotlagh It seems that `get_hotkey_for_net_and_neuron` and `get_neuron_for_net_and_hotkey` returns default 0 even if uid or hotkey doesn't exist. This could cause unexpected outcomes referring to an existing uid/key 0. Should those utils functions add an existence check first?

---

valid point! I changed functions to return error instead of default values.